exporter/metric/stdout/stdout.go

```diff
@@ -0,0 +1,187 @@
+	// basis.
+	Quantiles []float64
+}
+
+type expoBatch struct {
+	Timestamp *time.Time `json:"time,omitempty"`
+	Updates   []expoLine `json:"updates,omitempty"`
+}
+
+type expoLine struct {
+	Name      string      `json:"name"`
+	Max       interface{} `json:"max,omitempty"`
+	Sum       interface{} `json:"sum,omitempty"`
+	Count     interface{} `json:"count,omitempty"`
+	LastValue interface{} `json:"last,omitempty"`
+
+	Quantiles interface{} `json:"quantiles,omitempty"`
+
+	// Note: this is a pointer because omitempty doesn't work when time.IsZero()
+	Timestamp *time.Time `json:"time,omitempty"`
+}
+
+type expoQuantile struct {
+	Q interface{} `json:"q"`
+	V interface{} `json:"v"`
+}
+
+var _ export.Exporter = &Exporter{}
+
+func New(options Options) (*Exporter, error) {
+	if options.File == nil {
+		options.File = os.Stdout
+	}
+	if options.Quantiles == nil {
+		options.Quantiles = []float64{0.5, 0.9, 0.99}
+	} else {
+		for _, q := range options.Quantiles {
+			if q < 0 || q > 1 {
+				return nil, aggregator.ErrInvalidQuantile
+			}
+		}
+	}
+	return &Exporter{
+		options: options,
+	}, nil
+}
+
+func (e *Exporter) Export(_ context.Context, producer export.Producer) error {
+	// N.B. Only return one aggError, if any occur. They're likely
+	// to be duplicates of the same error.
+	var aggError error
+	var batch expoBatch
+	if !e.options.DoNotPrintTime {
+		ts := time.Now()
+		batch.Timestamp = &ts
+	}
+	producer.Foreach(func(record export.Record) {
+		desc := record.Descriptor()
+		labels := record.Labels()
+		agg := record.Aggregator()
+		kind := desc.NumberKind()
+
+		var expose expoLine
+		if msc, ok := agg.(aggregator.MaxSumCount); ok {
+			expose.Sum = msc.Sum().AsInterface(kind)
+			expose.Count = msc.Count()
+
+			if max, err := msc.Max(); err != nil {
+				aggError = err
+				expose.Max = "NaN"
+			} else {
+				expose.Max = max.AsInterface(kind)
+			}
+
+			if dist, ok := agg.(aggregator.Distribution); ok && len(e.options.Quantiles) != 0 {
+				summary := make([]expoQuantile, len(e.options.Quantiles))
+				expose.Quantiles = summary
+
+				for i, q := range e.options.Quantiles {
+					var vstr interface{}
+					if value, err := dist.Quantile(q); err != nil {
+						aggError = err
+						vstr = "NaN"
+					} else {
+						vstr = value.AsInterface(kind)
+					}
+					summary[i] = expoQuantile{
+						Q: q,
+						V: vstr,
+					}
+				}
+			}
+		} else if sum, ok := agg.(aggregator.Sum); ok {
+			expose.Sum = sum.Sum().AsInterface(kind)
+
+		} else if lv, ok := agg.(aggregator.LastValue); ok {
+			ts := lv.Timestamp()
+			expose.LastValue = lv.LastValue().AsInterface(kind)
+
+			if !e.options.DoNotPrintTime {
+				expose.Timestamp = &ts
+			}
+		}
+
+		var sb strings.Builder
+
+		sb.WriteString(desc.Name())
+
+		if labels.Len() > 0 {
+			sb.WriteRune('{')
+			sb.WriteString(labels.Encoded())
+			sb.WriteRune('}')
+		}
+
+		expose.Name = sb.String()
+
+		batch.Updates = append(batch.Updates, expose)
+	})
+
```

Comment:
nit: Shouldn't you return err here first if `aggErr != nil`?

Otherwise you might export output when there's an error that happened as you processed the batch?

Reply:
For this exporter specifically--if it is only an aggregator error not a marshal error--I figured it would be best to continue formatting the output and fill in "NaN". I figure NaN will parse as an error, but this at least yields a human-readable output. A single error is still returned, which the controller will print. I can be convinced to be less tolerant of errors though.